staging/src/k8s.io/apiserver/pkg/util/signal/signal.go

```diff
@@ -0,0 +1,63 @@
+package signal
+
+// TODO (p0lyn0mial): refactor lifecycle_signals.go to use this package
+
+// Interface encapsulates a named event
+type Interface interface {
+	// Signal signals the event, indicating that the event has occurred.
+	// Signal is idempotent, once signaled the event stays signaled and
+	// it immediately unblocks any goroutine waiting for this event.
+	Signal()
+	Receiver
+}
+
+// Receiver encapsulates read only methods
+type Receiver interface {
+	// Signaled returns a channel that is closed when the underlying event
+	// has been signaled. Successive calls to Signaled return the same value.
+	Signaled() <-chan struct{}
+
+	// Name returns the name of the signal, useful for logging.
+	Name() string
+
+	// IsSignaled is a convenience method for checking if the event has occurred.
+	IsSignaled() bool
+}
+
+func New(name string) Interface {
+	return &namedChannelWrapper{
+		name: name,
+		ch:   make(chan struct{}),
+	}
+}
+
+type namedChannelWrapper struct {
+	name string
+	ch   chan struct{}
+}
+
+func (e *namedChannelWrapper) Signal() {
+	select {
+	case <-e.ch:
+		// already closed, don't close again.
+	default:
+		close(e.ch)
```

Comment:
if `Signal` is used  concurrently, i think we need to use a `sync.Once` to protect it.